### ISSUE TITLE
Timeout during istio upgrade due to galley validation

### ISSUE THREAD

**Issue by Stono** [istio/istio#15091]
**Bug description**
Whilst upgrading a larger (508 services, 566 pods) cluster, the istio installation failed.  The tiller logs showed that it was waiting for a DestinationRule to be patched:

```
2019/06/24 13:32:45 Cannot patch DestinationRule: "disable-mtls-to-kube-dns" (Timeout: request did not complete within requested timeout 30s)
```

This is an old CRD that we have:
```
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: disable-mtls-to-kube-dns
  namespace: kube-system
spec:
  host: "kube-dns.kube-system.svc.cluster.local"
  trafficPolicy:
    tls:
      mode: DISABLE
```

We have done countless istio upgrades **without** galley - this upgrade **added** galley, so the only conclusion i can draw here is that it was waiting for galley, but hit the 30s timeout in the validatingwebhook configuration.

Note: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/

```
Note: Default timeout for a webhook call is 30 seconds but starting in kubernetes 1.14 you can set the timeout and it is encouraged to use a very small timeout for webhooks. If the webhook call times out, the request is handled according to the webhook’s failure policy.
```

**Note:** Our helm timeout is 5 minutes.

I cannot see anything in the galley logs about this DR:
```
❯ kubectl --namespace=istio-system logs -l istio=galley --since=10m | grep dns
```

Editing the resource directly a few minutes after the upgrade was fine:
```
❯ kubectl --namespace=istio-system edit destinationrule disable-mtls-to-kube-dns
destinationrule.networking.istio.io/disable-mtls-to-kube-dns edited
```

So my next working theory is that as we're adding galley to an existing cluster with a _lot_ of resources, Galley was churning its way through them.

However, galley doesn't look resource constrained at all:
<img width="923" alt="Screenshot 2019-06-24 at 14 37 33" src="https://user-images.githubusercontent.com/1486729/60023811-7b51ed80-968e-11e9-8a35-23dd2ba4bf8d.png">

This happened in two environments of a similar size.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[x] Installation
[ ] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastrcture

**Expected behavior**
Installations to work

**Steps to reproduce the bug**
You'd need to do an istio upgrade from 1.1.9 -> 1.2.0 on a pretty big cluster

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
Istio 1.2.0

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

**Comment by Stono** [comment]
For anyone else hitting this if you just want to disable galley validation for now `global.configValidation = false` will remove the `ValidatingWebhookConfiguration`

**Comment by Stono** [comment]
Extra info summarising from slack. Galley can only validate a single resource at any given time (makes sense, as it could be validating against other resources too, multiple validations brings in race conditions), however, when galley has a _lot_ of backed up changes, you can hit this timeout, for example - when galley starts up and it is validating all the crd's on the cluster.

In the context of this cluster, we have:

  - 417 Virtual Services 
  - 312 Destination Rules
  - 258 Sidecars

Therefore we have a pretty big period where we're unable to do any deployments of apps, or make any other changes after a galley restart.

**Comment by ozevren** [comment]
To clarify: my slack comment and this particular issue are two different things. My comment was about taking only one resource into consideration at a time, when checking the validity of a particular resource, and was within the context of an unrelated discussion.

**Comment by Stono** [comment]
I'm not trying to get into a "he said, she said" kind of debate - i'm simply trying to help you by finding issues.  Clearly we've gotten our wires crossed on the specifics so as I said on slack; I'll report here purely what I experienced which was a failure to deploy `1.2.0` due to a 30 second timeout on validation calls to galley and leave the rest up to you.

**Comment by ozevren** [comment]
Same here. My intention is to make sure we have the most relevant information in the report.

**Comment by ozevren** [comment]
And please do keep sending in bug reports.

**Comment by ozevren** [comment]
A few point-wise things: Galley validation webhook does not validate already-submitted resources. It only gets triggered by the API Server when there is an incoming CRUD request. The validation is also is very light on resource usage (as observed), so this is not a resource contention issue.

One possibility is a network hiccup (or a series of them) between the API Server and the hook causing the delays.

@Stono Do you have any other logs or K8s events that you can share, implicating the webhook & giving more details?

**Comment by ozevren** [comment]
Also, does your rollout update the control plane alongside the configurations, or are they applied separately?

**Comment by ozevren** [comment]
I ran some experiments locally using 1 replica, 10 replicas (disabled reconciliation) and no Galley (no webhook config), using a synthetic dataset based on the numbers provided here. The time taken for these operations are pretty close to each other. This is not entirely unexpected, as I expect the storage operation cost and throttling in API Server to be the dominating factor. 

```
1 Galley, create
real	2m27.771s
user	0m6.782s
sys	0m0.862s

1 Galley, delete
real	4m7.351s
user	0m6.206s
sys	0m0.922s

10 Galleys, create
real	2m29.431s
user	0m6.920s
sys	0m0.888s

10 Galleys, delete
real	4m5.567s
user	0m5.182s
sys	0m0.656s

No Galleys, create
real	2m22.541s
user	0m7.171s
sys	0m0.940s

No Galleys, delete
real	4m6.485s
user	0m5.557s
sys	0m0.704s

```

**Comment by Stono** [comment]
Hey thanks for the replies.

> One possibility is a network hiccup (or a series of them) between the API Server and the hook causing the delays.

Couldn't spot anything at the time, I checked the kubernetes event logs but there was nothing telling.

> Also, does your rollout update the control plane alongside the configurations, or are they applied separately?

It was a helm upgrade on the control plane, so at the same time as Galley coming up, pilot and mixer were also being restarted (1.1.9 -> 1.2.0).

> Do you have any other logs or K8s events that you can share, implicating the webhook & giving more details?

Nothing really, here's a graph of metrics from the kubernetes api tracking `histogram_quantile(0.95, sum(irate(apiserver_admission_webhook_admission_latencies_seconds_bucket{name="pilot.validation.istio.io"}[1m])) by (operation, type, le))` (please correct me if this isn't the right metric):
<img width="1663" alt="Screenshot 2019-06-25 at 07 48 22" src="https://user-images.githubusercontent.com/1486729/60075820-0f1eca80-971e-11e9-9b4e-6fa24220c436.png">

FYI the deployment/rollout started at 14:30 so you can see we were having problems for about 26 minutes.

As you can see there was a very, very minimal amount of requests going through:
<img width="1659" alt="Screenshot 2019-06-25 at 07 56 39" src="https://user-images.githubusercontent.com/1486729/60076173-d16e7180-971e-11e9-87d0-31a9d77acf40.png">

There were no rejected events recorded in the metrics.

**Comment by fpesce** [comment]
duplicate of #15088 , fixed by  #15254
